client/src/app/modules/components/sidebar/sidebar.component.ts

```diff
@@ -14,57 +14,88 @@ import { Course } from '@models/course'
-			// currently active/being worked in. 
-			const slug = `/courses/${course.id}`
-			if(url.indexOf(slug) == 0){
-				this.selectedCourse = course
-			}
-		})
-	}
-
-	// toggles what the currently selected course is
-	// this function is triggered by the UI element. 
-	handleToggleActiveCourse(course: Course){
-		if(this.selectedCourse && this.selectedCourse.id == course.id) this.selectedCourse = null
-		else this.selectedCourse = course
-	}
+  userCourses: Course[] = []
+  selectedCourse: Course = null
+  activeLink: number;
+  links=["home", "sections", "problems", "settings"];
 
+  constructor(
+    private _userService: UserService,
+    private _courseService: CourseService,
+    private router: Router,
+  ) {
+  }
+
+  ngOnInit() {
+    this._userService.getCurrentUser().subscribe({
+      next: async (user) => {
+        this.userCourses = await this._courseService.getCourses()
+        this.setActiveCourseFromRouter()
+      }
+    })
+
+    // this router subscription will follow any changes in the current route
+    // this will allow for us to determine which course is currently being focused on,
+    // which subpiece to highlight, etc.
+    // we do this by examining URL changes, and on each change, breaking the URL down into its slugs and
+    // comparing those slugs with what we current have stored.
+    this.router.events.subscribe({
+      next: async (event) => {
+        // ensure this event describes a URL change
+        if (event && event["url"]) {
+          this.setActiveCourseFromRouter();
+          // this.setActiveNavLinkFromRouter();
+        }
+      }
+    })
+  }
+
+  // this function will check what's going on in the router,
+  // and attempt to set the active course in the sidebar depending on what the loaded course is
+  setActiveCourseFromRouter() {
+    const url = this.router.url;
+    this.userCourses.forEach((course) => {
+      // this slug is the prefix to the URL that would indicate that this course is
+      // currently active/being worked in.
+      const slug = `/courses/${course.id}`;
+      if (url.indexOf(slug) == 0) {
+        this.selectedCourse = course;
+        // console.log("selected course: " + this.selectedCourse.name);
+        this.setStep(this.userCourses.indexOf(this.selectedCourse))
+      }
+    });
+    this.setActiveNavLinkFromRouter();
+  }
+
+  setActiveNavLinkFromRouter() {
+    const url = this.router.url;
+    if (this.selectedCourse) {
+      const slug = `/courses/${this.selectedCourse.id}`;
+      const page = url.substr(slug.length+1);
+
+      this.activeLink=this.links.indexOf(page);
+      if(page==""){
+        this.activeLink=0;
+      }
+      // console.log('page: ' + this.links[this.activeLink]);
+
+    }
+  }
+
+  // toggles what the currently selected course is
+  // this function is triggered by the UI element.
+  handleToggleActiveCourse(course: Course) {
+    // if (this.selectedCourse && this.selectedCourse.id == course.id) this.selectedCourse = null
```

Comment:
if this code isn't needed, please delete it. If the function isn't needed, feel free to delete it likewise.